confluent_kafka/producer.py

```diff
@@ -0,0 +1,139 @@
+        :raises TypeError: If conf is not a dict object.
+    """
+
+    def __new__(cls, *args, **kwargs):
+        if 'key_serializer' in kwargs or 'value_serializer' in kwargs:
+            return super(Producer, cls).__new__(SerializingProducer, *args, **kwargs)
+        return super(Producer, cls).__new__(cls, *args, **kwargs)
+
+
+class SerializingProducer(Producer):
+    """
+        SerializingProducer extends Producer with configurable key and value serializer.
+
+        To avoid spontaneous calls from non-Python threads all callbacks will only be served upon
+            calling ```client.poll()``` or ```client.flush()```.
+
+        Instances of SerializingProducer cannot be created directly.
+        To obtain an instance of this class instantiate a Consumer with a key and/or value deserializer.
+
+        Duplicate params have been omitted for brevity. See Consumer for class documentation.
+
+        :raises TypeError: If conf is not a dict.
+        :raises TypeError: If instantiated directly.
+    """
+
+    __slots__ = ["_key_serializer", "_value_serializer"]
+
+    def __new__(cls, *args, **kwargs):
+        raise TypeError("SerializingProducer is a non user-instantiable class")
+
+    @staticmethod
+    def byteSerializer(topic, data):
+        """ Pass-through serializer """
+        return data
+
+    # conf must remain optional as long as kwargs are supported
+    def __init__(self, conf={}, key_serializer=None, value_serializer=None,
+                 error_cb=None, stats_cb=None, throttle_cb=None, logger=None, **kwargs):
+
+        if not isinstance(conf, dict):
+            raise TypeError("expected configuration dict")
+
+        if kwargs:
+            # Handle kwargs for backwards compatibility
+            conf.update(kwargs)
+            warn("The use of kwargs is being deprecated. "
+                 "In future releases `conf` will be mandatory and "
+                 "all keyword arguments must match the constructor signature explicitly.",
+                 category=DeprecationWarning, stacklevel=2)
+
+        # Ensure the default serializer cannot be overwritten with None on instantiation
+        if key_serializer is None:
+            key_serializer = SerializingProducer.byteSerializer
+
+        if value_serializer is None:
+            value_serializer = SerializingProducer.byteSerializer
+
+        self._key_serializer = key_serializer
+        self._value_serializer = value_serializer
+
+        # Callbacks can be set in the conf dict or *ideally* as parameters.
+        # Raise a SyntaxError if a callback is set in both places.
+        for var, name in [(logger, 'logger'), (error_cb, 'error_cb'),
+                          (stats_cb, 'stats_cb'), (throttle_cb, 'throttle_cb')]:
+            if all([var, conf.get(name, None)]):
+                raise SyntaxError("{} parameter repeated".format(name))
+            if var is None:
+                var = conf.get(name, None)
+
```

Comment:
Unless here I add
```
conf.pop("key_serializer")
conf.pop("value_serializer")
```

instance creation fails with 
```
cimpl.KafkaException: KafkaError{code=_INVALID_ARG,val=-186,str="No such configuration property: "key_serializer""}
```

I think `librdkafka` does not accept these properties, I tried with both `1.0.0rc5` and `master`.